### ISSUE TITLE
Multi-Machine Parallel Builds (fan-in, fan-out)

### ISSUE THREAD

**Issue by JeanMertz** [drone/drone#1814]
I'm currently running Drone 0.5 on Kubernetes for our company. I read about the Matrix feature, but what I haven't found yet is if there's a fanout/fanin kind of system in place.

Say I want to do this during a run:
- build binary on one worker
- fanout to five workers, running tests with that binary
- fanin to one worker to deploy / notify / ...

There are other use-cases here that we'd love to use – mostly involving speeding up test runs through parallelisation – all depending on a fanout/in system.

Is this possible, or is there an issue tracking a use-case like this?

**Comment by bradrydzewski** [comment]
We have a specification for single-machine fan-in / fan-out at https://github.com/drone/drone/issues/1356

With single-machine fan-in / fan-out perhaps you could scale vertically? For example you could provision a Google Computes n1-highmem-32 (32 virtual CPUs and 208 GB of memory) for single-machine parallelism. I'm sure this could yield some nice performance improvements.

There are two challenges with multi-machine fan-in / fan-out support. First is that drone expects a shared workspace (volume) across all steps in the build process. The second challenge is granularity of scheduling. Builds are scheduled per-machine as opposed to scheduling steps per-machine.

Perhaps a good workaround would be to create a custom plugin that handles the fan-in / fan-out to multiple machines. The plugin could (for example) use the docker remote API to launch tests across multiple machines and aggregate their results ...

**Comment by JeanMertz** [comment]
Interesting.

To be honest, within our company we are moving everything on top of Kubernetes right now. Within Kubernetes, we heavily rely on multiple machines with semi-high specs, but not the kind of specs you are talking about. Instead, we horizontally scale nodes when needed (and will soon auto-scale horizontally).

Currently, we have a 4-node drone instance running:

```
$ kubectl get pod --namespace=drone
NAME                      READY     STATUS    RESTARTS   AGE
agent-47ont               1/1       Running   0          1d
agent-9ha8c               1/1       Running   0          1d
agent-rfejq               1/1       Running   0          1d
agent-tyo78               1/1       Running   0          1d
server-1499183458-7suy4   1/1       Running   0          1d
```

which would unfortunately not work with the single-machine solution.

I can see the challenges here, and am not providing any solutions at this point in time, but hopefully our usecase gives you some insight in how customers are using your product, and why it might make sense (imprinting our use-case to all of your customers right now, sorry) to not focus too heavily on single-machine setups, given how we're seeing a lot of partners/teams slowly moving up the stack from Docker to Kubernetes.

**Comment by bradrydzewski** [comment]
> To be honest, within our company we are moving everything on top of Kubernetes right now. Within Kubernetes, we heavily rely on multiple machines with semi-high specs, but not the kind of specs you are talking about. Instead, we horizontally scale nodes when needed (and will soon auto-scale horizontally).

couldn't you have a set of reserved machines for your CI process? You could put drone agent on a subset of machines (0.5 supports agents which is much more k8s friendly) and you could designate those machines accordingly.

> which would unfortunately not work with the single-machine solution.

just to clarify, I'm not proposing only using 1 machine for drone. You can use many (hundreds) of machines with drone, but a build can only be executed on a single host. You could fan-out the build steps on that single host for parallelism, as described in #1356 

> and why it might make sense (imprinting our use-case to all of your customers right now, sorry) to not focus too heavily on single-machine setups

see note above :)

**Comment by bradrydzewski** [comment]
> but hopefully our usecase gives you some insight in how customers are using your product

side note, I'm not sure "customers" is the appropriate term here, since nobody pays me :)

**Comment by JeanMertz** [comment]
> 0.5 supports agents which is much more k8s friendly

Yes, the above output is running 0.5 which works flawlessly on Kubernetes.

One of the reasons by the way why I created https://github.com/drone/drone/issues/1815 was that Drone takes direct control of the Docker daemon, which circumvents scheduling and resource rules set in place by Kubernetes, which makes it dangerous to run Drone on the "main" cluster. For now, I've set up a smaller side-cluster, to cope with this.

> couldn't you have a set of reserved machines for your CI process

We can indeed. We can create "node-pools", and assign drone agents to pools with specific labels (like high-mem,high-cpu).

> I'm not proposing only using 1 machine for drone. You can use many (hundreds) of machines with drone, but a build can only be executed on a single host. You could fan-out the build steps on that single host for parallelism

Ah, that clarifies a lot. I guess when you want to go extreme, parallelising across machines makes sense, but for a matrix of 5 to 10 jobs, having those be confined to a single machine is perfectly feasible, especially if other runs can then do the same on other machines within the cluster.

That sounds really good. I'll go subscribe to that issue and follow it along 😄

> I'm not sure "customers" is the appropriate term here, since nobody pays me

Yeah, sorry, didn't mean to put it like that. Although I'm sure you have _some_ kind of monetization strategy in mind, right? Such an awesome tool, so much commitment (I've been following along on and off since 0.2), I hope it pays off for you in the end 👍

**Comment by tboerger** [comment]
I already suggested to write a custom agent that uses the k8s api instead of the docker api. That should be pretty neat.

**Comment by mwotton** [comment]
as a data point, we are currently leaving drone for lack of this feature.

**Comment by minhchuduc** [comment]
@JeanMertz i curious about which setup you're using to run drone-agent on Kubernetes. Are you using sibling containers or DinD setup?

**Comment by bradrydzewski** [comment]
implemented in 1.0.0-rc.1
https://readme.drone.io/config/pipeline/multi-machine/
